M20 S L options

RFBomb opened the pull request:
### Requirements

Serial Communication (such as octoprint) required to test functionality. 

### Description

Updated files for expanded Long_FileName support

/src/gcode/sd/M20.cpp
- Added L and S arguments.
- Added ability to specify preference if no L / S argument specified.

/src/gcode/sd/M33.cpp
- Modified command that requests the feedback to work with the reworked routine.

/Configuration_adv.h
- Added two new defines to trigger the defaults for M20 rework

/src/sd/cardreader.h
- Changes necessary to work with reworked cardreadper.cpp file

/src/sd/cardreader.cpp
- Expanded functionality for Long FileNames
- Modified SelectByName routine to be able to return the file if a long name was specified instead of the 8.3 name
- Modified Card.LS function to work with new M20 functionality
- Modified PrintListing function to work with new M20 Functionality
    - - If requesting long filenames back, this calls the 'printLongPath' function to retrieve said filenames
- Modified printLongPath function to be compatible with new PrintListing routine functionality
- Modified diveToFile routine to be able to support Long FileNames
- This routine is what most others use to pick and choose files on the card for reading and writing.
- Adding support in this routine automatically brought the rest up to par.
- Routine works with both DOS8.3 names and long names now

**Known Issues:**

1.
- When uploading a file to SD with octoprint, it was uploaded using the 8.3 format (expected).
- Calling M33 to read the long filename of this file resulted in the first file with a long filename to show up instead. For me it was firmware.bin that was being displayed instead of 'PLA_LI~1.GCO'
- After inspecting the SD card itself, the 8.3 format was the actual filename octoprint declared to upload. 
- I have put in a string comparison check in order to prevent this issue from occuring though. The workaround will display the 8.3 filename if the first 3 characters of the long filename don't match up with the DOS name.
- It would be up to hosts to implement long filename uploading support on their end. 

2.

- Writing to the SD card using M28/M29 still requires use of the 8.3 filename. (The rest of the path should work in long form though, as long as the folder path exists.
- This is due to how SdBaseFile.cpp is written.
- This module is expecting the 8.3 format, and if it isn't give then filename in that format it rejects the operation.
- I'm fairly confident it has to do with this command line in the ' ::open( ' routine:
`    		if (!make83Name(path, dname, &path)) return false;`


### Benefits

This is a long requested feature by many users. 
This is also a contested feature by several hosts that refuse to call M33 in the background due to fear of overloading the serial communication. Those hosts (such as octoprint) would rather a firmware solution. This provides that in a customizable format. 

It also is able to accept long form filenames for selecting files / deleting files. As noted above, writing new files needs other work to be compatible. Octoprint is supposedly going to have support for long file names in v1.5 which is upcoming, and its up to firmware to make it compatible on the printer side. 

### Configurations

Several new defines have been added in Configuration_adv.h to mess with the M20 defaults if L/S is not specified. 
Other than that, its simply enabling long file support and using M20 L / M20 S  / M20 LS. 

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/19960
https://github.com/MarlinFirmware/Marlin/pull/19111

RFBomb committed:
Expanded Long_FileName Support

Updated files for expanded Long_FileName support
     - /src/gcode/sd/M20.cpp
    	- Added L and S arguments.
    	- Added ability to specify preference if no L / S argument specified.
    - /src/gcode/sd/M33.cpp
    	- Modified command that requests the feedback to work with the reworked routine.
    - /Configuration_adv.h
    	- Added two new defines to trigger the defaults for M20 rework
    -  /src/sd/cardreader.h
    	- Changes necessary to work with reworked cardreadper.cpp file
    -  /src/sd/cardreader.cpp
    	- Expanded functionality for Long FileNames
    	- Modified SelectByName routine to be able to return the file if a long name was specified instead of the 8.3 name
    	- Modified Card.LS function to work with new M20 functionality
    	- Modified PrintListing function to work with new M20 Functionality
    		- If requesting long filenames back, this calls the 'printLongPath' function to retrieve said filenames
    	- Modified printLongPath function to be compatible with new PrintListing routine functionality
    	- Modified diveToFile routine to be able to support Long FileNames
    		- This routine is what most others use to pick and choose files on the card for reading and writing.
    		- Adding support in this routine automatically brought the rest up to par.
    		- Routine works with both DOS8.3 names and long names now.

    #Known Issues:
    - When uploading a file to SD with octoprint, it was uploaded using the 8.3 format (expected).
     	- Calling M33 to read the long filename of this file resulted in the first file with a long filename to show up instead. For me it was firmware.bin that was being displayed instead of 'PLA_LI~1.GCO'
    	- After inspecting the SD card itself, I don't know if theres a workaround for this yet.
    	- I have put in a string comparison check in order to prevent this issue from occuring though. The workaround will display the 8.3 filename if the first 3 characters of the long filename don't match up with the DOS name.

    - Writing to the SD card using M28/M29 still requires use of the 8.3 filename. (The rest of the path should work in long form though, as long as the folder path exists.
    	- This is due to how SdBaseFile.cpp is written.
    	- This module is expecting the 8.3 format, and if it isn't give then filename in that format it rejects the operation.
    	- I'm fairly confident it has to do with this command line in the ' ::open( ' routine:
    		if (!make83Name(path, dname, &path)) return false;

thinkyhead commented:
You should come over to the Marlin Discord server to discuss this stuff. We already have a series of proposals under development, with open PRs, that have been needing attention from someone interested in the subject. Join link: https://discord.gg/n5NJ59y

thinkyhead committed:
Pretty up, modernize, etc.

thinkyhead committed:
Reverse reversal

thinkyhead committed:
Use PGM strings

ellensp commented:
Only LPC1768/9 framework seems to have TRUE and FALSE, rest use true and false.

thinkyhead committed:
More cleanup

thinkyhead committed:
Change M20 short name option

thinkyhead committed:
Guess not…

rhapsodyv commented on Marlin/src/sd/cardreader.cpp:
```diff
@@ -222,20 +223,110 @@ void CardReader::selectByIndex(SdFile dir, const uint8_t index) {
 // Get file/folder info for an item by name
 //
-void CardReader::selectByName(SdFile dir, const char * const match) {
+void CardReader::selectByName(SdFile dir, const char * const match, const bool debug/*=false*/) {
```
runtime debug?

Looking at this code, all this debug code seems to be only useful while developing. After that, I don't think it's needed.

RFBomb replied:
Yes, that was part of a runtime debugger I had set up. 
The reason for the parameter in the routine like that was so that while I was debugging i could choose to display the debugger items in this routine for problematic operations, but not for ones that work. IE: I knew it was working fine with M20's PrintListing, but was having issues with the DiveToFile routine.  Those actual debugging lines would get stripped out of the code during compile time (except for the debug boolean being set to false) if the main debugger for that module was not defined.

Edit: well it was set up to be stripped out between lines 249-253, but the `#if(Enabled` has been replaced with    `if (TERN0(DEBUG_CARDREADER`, which I'm not sure what that does. (Seems to be an 'AND' but I've never seen that function)

RFBomb commented:
@thinkyhead  I had seen some of the other discussions that have taken place on various forums regarding the topic, and was unaware of any work that had been done (from what I could tell it was mostly people being confused as to the 'why cant it' or people requesting others to implement it.)  I hadn't seen other pull requests with it (maybe I missed them), so I decided to build it in since I wanted to better manage my printer's sd card anyway. 

I took a look at the mods youve done to what I uploaded, and its quite a bit cleaner. (I typically program in VBA more than other languages, so I had to refresh myself on how C++ works during this code writing). 

I believe I tracked down the last major piece of the puzzle to the `::open(` routines inside of `SdBaseFile.cpp`, as noted in the initial pull request. If the filename/path provided isn't in DOS8.3 format, these routines just exit the routine. Because there are 3 overrides for that, I'm hesitant to touch them (that and its close enough to the SD card that I don't want to break anything, since its likely used a lot more than I'm anticipating). But I believe these functions are likely the key to getting M28 + M29 working. M28 is using the DiveToFile routine to select the path writing just fine using the long file name, but since its not in 8.3 format, the file cannot be opened for writing due to these `open ` routines. 

Note that I only examined the routines that explicitly deal with the SD card via M2X/M3X gcode. Any of the other functions that may utilize other storage devices haven't been looked at, but (asuming they are coded similar), the changes could probably be relatively easily be brought to those cpp files I imagine. 

Also, as rhapsodyv noted, once we know everything is working fine (it was from my brief testing), that debug code should be able to come out. I was just using it for selective debugging based on what was being tested between compiles. I figured I'd leave it in there to allow other people testing the changes out to have some way to view progress/what the cpu is doing during while testing the routines.

thinkyhead commented:
Reading of the long filename element of a directory entry was hacked into the older version of `SdFile` embedded in Marlin quite a long time ago, but there is no support for creating long filename directory entries in our embedded `SdFile`.

I've been recently using the Marlin simulator, which now has a simulated SD card, to start exploring good augmentations for `M20` which will satisfy the needs of host developers. I've also started exploring extensions to the file listing stuff in `CardReader` to interoperate better with the PanelDue controller, which speaks JSON. I'm taking this cautiously step by step, because I want to optimize these processes and avoid "thrashing" the media. I've been working out these extensions over at #10610 and #19111. Since the latter PR is now closed, I will move that experimentation over to the earlier one.

RFBomb commented:
so the plan is to likely move to returning a json object over serial to the host? Or would this be one of the options a user could use? I think delivering the json package would definitely make the job of any host's developer's lives easier to work with. 

I took a peek over at #10610 's cardreader.cpp to investigate it what you guys are working on over there. I think some of what I've written could be useful to integrate into it maybe. Obviously you guys that are actively working on it have a much better idea of the current state + intentions, but purely looking at the `CardReader::diveToFile` and the `CardReader::selectByName` routines leads me to think they still only accept 8.3 formats as inputs. The changes I've put into this pull for those routines should allow them to dive/select paths given in the long format. Of course, if the intention is that the JSON package returns both of those (pretty path for user comfort, but 8.3 path to speak to the printer) then this addition would be unnecessary.

thinkyhead commented:
Some host developers don't like JSON, so for those developers I am looking at simplifying the file listings and also providing both hosts and firmware developers a simple way to traverse the directory tree, similar to a shell.

The first consideration is that we still love DOS8.3 paths … but only because they are compact and unique. The most-compatible `M23` will always takes the full DOS path to the G-code file to be opened for printing, and that doesn't need to change. But it would definitely be useful to get folder- and file-selection by long-name working in `M23` as well. There is now quoted string support in Marlin, so this should be required in order to use long names and long paths.

Since it's a longstanding issue that `M20` is redundant and can definitely be shrunk down, I'm working on making it more compact and eliminating all redundancy. JSON will be fine and is needed for PanelDue, but something that looks like YAML may work better for our host developer friends….

thinkyhead committed:
ExtUI homing / leveling additions

sjasonsmith commented:
@thinkyhead were there also recent improvements that make this PR obsolete, or is this still an area for ongoing development?

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19977

thinkyhead commented:
@sjasonsmith — I am continuing to experiment with alternative listing methods, including in the "PanelDue" PR which has its own idea about how to get listings in JSON format. In general, I need to continue to clean up, organize, and document the CardReader class, and maybe separate out the "file browser" (which bolts the concept of "workDir" onto the CardReader class and assumes a single CWD).

thinkyhead committed:
[cron] Bump distribution date (2020-11-26)

swissnorp committed:
Probe Offset Wizard improvements (#20239)

thinkyhead committed:
Reduce warnings, extern "C" cleanup (#20279)

thinkyhead committed:
Optimize emergency parser check

thinkyhead committed:
No auto debug for EEPROM_CHITCHAT

sjasonsmith committed:
Retire HAL for STM32F4 / F7 (#20153)

thinkyhead committed:
Tweak some pin errors

mks-viva committed:
MKS Robin E3 / E3D v1.1 (#20216)

thinkyhead committed:
Finish HAL/STM32 cpp wrappers

wmariz committed:
Level Corners with Probe option (#20241)

daleckystepan committed:
[WIP] Fix Probe::offset_xy (#20290)

Co-authored-by: Jason Smith <jason.inet@gmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-11-27)

thinkyhead committed:
M808 Repeat Markers (#20084)

BsCmOD committed:
FIX TMC menu message (#20294)

ellensp committed:
Fix BTT GTR 1.0 endstop/DIAG pins (#20296)

Commit:
Fix G2/G3 arcs > 180° (#20292)

Sergey1560 committed:
Group related homing options (#20283)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

niccoreyes committed:
Fix M73 LCD code typo (#20300)

thinkyhead committed:
[cron] Bump distribution date (2020-11-28)

thinkyhead committed:
[cron] Bump distribution date (2020-11-29)

Commit:
Arc Direction followup for circles (#20314)

kad committed:
SKR E3 Turbo Controller Fan (#20320)

sjasonsmith committed:
Fix STM32F1 'freeMemory()' warnings (#20319)

sjasonsmith committed:
Leveling Fade Height default setting (#20316)

thinkyhead committed:
[cron] Bump distribution date (2020-11-30)

thinkyhead committed:
Adjust axis homed / trusted methods (#20323)

thinkyhead committed:
General cleanup

sjasonsmith committed:
Increase E3V2 DWIN steps/mm range to 999.9 (#20324)

thinkyhead committed:
[cron] Bump distribution date (2020-12-01)

sjasonsmith committed:
Prevent Watchdog reset writing Creality 4.x EEPROM (#20328)

sjasonsmith committed:
Fix E3V2 Control Menu when returning from Info (#20338)

thinkyhead committed:
Tweak to EEPROM safe delay

thinkyhead committed:
[cron] Bump distribution date (2020-12-02)

thinkyhead committed:
Level Bed Corners is a sub-menu

mathiasvr committed:
Update to STM32 v10, optimize build (#20325)

shitcreek committed:
Help hosts when password-locked (#20348)

sjasonsmith committed:
Fix E3V2 DWIN Jerk Menu (#20352)

* Fix E3V2 DWIN build without CLASSIC_JERK
* Fix jerk edits applying to wrong index

sjasonsmith committed:
Permit SD EEPROM emulation on E3V2 (#20353)

thinkyhead committed:
[cron] Bump distribution date (2020-12-03)

Commit:
Fix circle arc condition (#20322)

ellensp committed:
Fix up start, monitor baud (#20326)

thinkyhead committed:
Consolidate common pin includes

rhapsodyv committed:
Require minimum PlatformIO version (#20361)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>
Co-authored-by: Jason Smith <jason.inet@gmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-04)

sjasonsmith committed:
Fix MESH_BED_LEVELING w/o SEGMENT_LEVELED_MOVES (#20363)

thebelin committed:
Loosen E on pause for fila-manipulation (#20346)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-05)

tpruvot committed:
Don’t use near keyword as variable name (#20374)

thinkyhead committed:
Improved Longer3D LKx Pro board (#20372)

Co-authored-by: mrv96 <marione96@live.it>

sjasonsmith committed:
Fix TMC_HOME_PHASE divide by zero (#20368)

sjasonsmith committed:
Suspend Servos for STM32+NeoPixel (#19963)

InsanityAutomation committed:
Add Creality 4.5.2 board (#20378)

Co-authored-by: Jason Smith <jason.inet@gmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-06)

thinkyhead committed:
Creality 4.5.2 format and fix

LinFor committed:
Fix FYSETC S6 I2C EEPROM size (#20340)

Both V1.2 and V2.0 boards have a 24LC16, which is a 2kB EEPROM.

Co-authored-by: PingWin <kirill@shashlov.ru>
Co-authored-by: Jason Smith <jason.inet@gmail.com>

thinkyhead committed:
Use 'nearby' for 'near'

thinkyhead committed:
G28 tweaks

thinkyhead committed:
Fix STM32F1 SERIAL_GET_TX_BUFFER_FREE

thinkyhead committed:
Preheat before leveling

thinkyhead committed:
Move G29 3-point startup earlier

thinkyhead committed:
Always enable leveling after G28

thinkyhead committed:
G28 followup

rhapsodyv committed:
Avoid invalid memory optimizations (#20389)

When building for AVR, merge-all-constants can incorrectly combine constants stored in flash with constants stored in RAM. These have different access requirements, leading to undefined behavior during execution.
Co-authored-by: ellensp <ellensp@hotmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-07)

thinkyhead committed:
Watch idle() depth over 5

thinkyhead committed:
Fix fileExists, use openFailed

thinkyhead committed:
More LPC P-string macros

thinkyhead committed:
Fix auto#.g file handling, add NO_SD_AUTOSTART (#20071)

ellensp committed:
Shorten Filament Load/Unload Strings to Fit on Graphical Displays (#20369)

Co-authored-by: ellensp <ellensp@ellensp-HP-ProBook-6470b.fritz.box>
Co-authored-by: Victor Mateus Oliveira <rhapsodyv@gmail.com>

ellensp committed:
Allow BTT EXP-MOT be used with Displays that needs only EXP 1 (#20396)

Co-authored-by: ellensp <ellensp@ellensp-HP-ProBook-6470b.fritz.box>
Co-authored-by: Victor Mateus Oliveira <rhapsodyv@gmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-08)

rhapsodyv committed:
SPI TFT for STM32F4 boards (#20384)

* fix pinsDebug for F1 boards

* add MKS Robin PRO V2 board - development board

* tft spi working with F4 boards

* pins formating

* sanity check for TFT on supported cores in STM32

* Fix tabs/spaces in pins file

Co-authored-by: Jason Smith <jason.inet@gmail.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-09)

thinkyhead committed:
Style, spacing, typo cleanup for recent changes

thinkyhead committed:
Tweaks for pins_BTT_SKR_common.h

thisiskeithb committed:
Clarify Delta & SCARA config location (#20403)

swissnorp committed:
Allow positive Z nozzle to probe offset (#20344)

* Allow a positive value for z offset from nozzle to probe without sanity checks
* Simplify menu_probe_offset.cpp (HOMING_Z_WITH_PROBE)
* Add some more explanation to Configuration Files
* Raise after probe_at_point as if homed

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>
Co-authored-by: Jason Smith <jason.inet@gmail.com>

descipher committed:
Fix Spindle/Laser Control menu (#20347)

marciot committed:
Fixes and additions to FTDI EVE Touch UI (#20393)

- Fixed name conflict with "SUBSCRIPT_TWO"
- Fixed rendering bugs in "Leveling Menu"
- Only show "Bed Mesh Screen" when UBL is enabled
- Removed CocoaPress code from generic "Main Menu"
- Removed CocoaPress code from LulzBot Bio "Status" screen
- Moved generic "Move Axis" functionality into a base class
- Added CocoaPress custom screens:
   - Status Screen
   - Unload Cartridge
   - Load Chocolate
   - Main Menu
   - Advanced Settings
   - XYZ Move (based on "Move Axis" base class)
   - Extrusion Move (based on "Move Axis" base class)
- CocoaPress tweaks to "Temperature" screen
- Fix FTDI EVE Touch UI compilation errors when not using leveling.

InsanityAutomation committed:
Move ExtUI ABL mesh edit, limit to bilinear (#20381)

* Move ExtUI call

Co-authored-by: Scott Lahteine <github@thinkyhead.com>
Co-authored-by: Victor Mateus Oliveira <rhapsodyv@gmail.com>
Co-authored-by: Sebastiaan Dammann <sebastiaandammann@outlook.com>

rhapsodyv committed:
MKS Robin Pins fixes for STM32 and STM32F1 (#20404)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-10)

thisiskeithb committed:
Clarify "not Interrupt-capable" error message (#20419)

ellensp committed:
Fix UBL Debug Messages (#20423)

Co-authored-by: ellensp <ellensp@hotmsil.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-11)

elasticdotventures committed:
Z_SENSORLESS sanity checks (#20421)

Co-authored-by: Jason Smith <jason.inet@gmail.com>

FanDjango committed:
Probe offset wizard fixes (#20414)

* STOW probe, reverting incorrect earlier change from #20344
* Adjust soft endstop disables, to ensure travel below bed functions properly

Co-authored-by: FanDjango <FanDjango@users.noreply.github.com>

sjasonsmith committed:
SENSORLESS_PROBING sanity check followup (#20438)

thinkyhead committed:
[cron] Bump distribution date (2020-12-12)

thinkyhead committed:
Better animated boot screen

leodoener committed:
Fix SDCARD_SORT_ALPHA on Ender 3 V2 (#20443)

ellensp committed:
Add HAS_PIN_27_BOARD for CREALITY_V4 (#20446)

Co-authored-by: ellensp <ellensp@hotmsil.com>

qwewer0 committed:
Improve ASSISTED_TRAMMING_WIZARD probe stowing (#20437)

* When BLTOUCH_HS_MODE enabled, stow pin before user interaction
* For all probes, ensure probe stows at end of wizard

thinkyhead committed:
[cron] Bump distribution date (2020-12-13)

FanDjango committed:
Probe Offset Wizard followupBack to PROBE_PT_RAISE/separate STOW, make "PROBING" msg appear (#20439)

* Go back to always use PROBE_PT_RAISE with a discrete stow. This ensures a raise above the bed, while stowing prior to exiting the wizard.
* Fix issue preventing text while moving to X/Y position
Co-authored-by: FanDjango <FanDjango@users.noreply.github.com>

thinkyhead committed:
Robin pins followup

thinkyhead committed:
SHORT_MANUAL_Z_MOVE => FINE_MANUAL_MOVE

thinkyhead committed:
Adjust planner debugging

thinkyhead committed:
[cron] Bump distribution date (2020-12-14)

Sebazzz committed:
Creality v4.5.3 (CR-6 SE, CR-6 MAX) (#20468)

p3p committed:
LPC176x framework update (#20469)

thinkyhead committed:
[cron] Bump distribution date (2020-12-15)

thinkyhead committed:
Fix formatting

sjasonsmith committed:
Trust that script -x flags will always be set (#20453)

kisslorand committed:
Fix M1001 auto-check logic (#20456)

thinkyhead committed:
Auto-check followup

thinkyhead committed:
Please install the EditorConfig plugin

GadgetAngel committed:
Fix RESTORE_LEVELING_AFTER_G28 (#20471)

thinkyhead committed:
Update some pins errors

thinkyhead committed:
Trust STM32 gcc versions

thinkyhead committed:
Update some LCD includes

thinkyhead committed:
[cron] Bump distribution date (2020-12-16)

GMagician committed:
Update Italian language (#20480)

thinkyhead committed:
Update Hungarian language

Co-Authored-By: Zs.Antal <45710979+AntoszHUN@users.noreply.github.com>

tpruvot committed:
Update French language (#20472)

thinkyhead committed:
[cron] Bump distribution date (2020-12-17)

thinkyhead committed:
Use homing_feedrate function

rafaljot committed:
Homing feedrates as XYZ array (#20426)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

ellensp committed:
Fix ZoneStar LCD 2004 buttons (#20489)

nb-rapidia committed:
Make M220 B / R a standard feature (#20355)

sjasonsmith committed:
Balance CI tests (#20485)

grauerfuchs committed:
Fix and optimize MightyBoard (#20493)

thinkyhead committed:
[cron] Bump distribution date (2020-12-18)

rhapsodyv committed:
MKS Robin Nano V3 and STM32F4x0Vx Variant (#20430)

tpruvot committed:
DOGM: Slow down touchscreen calibration (#20454)

thinkyhead committed:
[cron] Bump distribution date (2020-12-19)

thinkyhead committed:
Rename QUIET_PROBING

thinkyhead committed:
[cron] Bump distribution date (2020-12-20)

thisiskeithb committed:
Let boards set Default TMC Slave Addresses (#20498)

Co-authored-by: Scott Lahteine <github@thinkyhead.com>

InsanityAutomation committed:
Minimum temp options for Probing and G12 Nozzle Clean (#20383)

Co-authored-by: Jason Smith <jason.inet@gmail.com>
Co-authored-by: Scott Lahteine <github@thinkyhead.com>

InsanityAutomation committed:
Probe Tare, Probe Activation Switch (#20379)

Co-authored-by: Scott Lahteine <github@thinkyhead.com>
Co-authored-by: Victor Mateus Oliveira <rhapsodyv@gmail.com>
Co-authored-by: Jason Smith <jason.inet@gmail.com>

sjasonsmith committed:
Fix Creality EEPROM watchdog freq. (#20510)

Foxies-CSTL committed:
Migrate Hispeedv1 (QQS-Pro) to HAL/STM32 (#20354)

thinkyhead committed:
Languages cleanup

Serhiy-K committed:
Update Russian and Ukrainian languages (#20508)

thinkyhead committed:
Fix PIO typo

Sebazzz committed:
Add BTT SKR CR6 board (#20522)

thinkyhead committed:
[cron] Bump distribution date (2020-12-21)

streef committed:
Get E3V2 DWIN `MACHINE_SIZE` from config (#20526)

streef committed:
Apply HOME_AFTER_DEACTIVATE for 'G28 O' (#20525)

thinkyhead committed:
Clean up some pins

Commit:
Fix G34 compile with bed leveling disabled (#20537)

LinFor committed:
Fix some Russian (#20529)

thinkyhead committed:
Tweak FTDI spacing

LinFor committed:
Add Tune -> Advanced Settings to FTDI EVE (#20532)

thinkyhead committed:
[cron] Bump distribution date (2020-12-22)

LinFor committed:
12-bit ADC resolution for STM32 (#20519)

LinFor committed:
FTDI EVE custom user menus (#20518)

sjasonsmith committed:
Improve STM32 timer conflict messages (#20544)

p3p committed:
Fix UBL mesh inset Z position (#20538)

sjasonsmith committed:
Overrides to prevent STM32 timer conflicts (#20545)

X-Ryl669 committed:
Improve Touch Calibration screen (#20524)

smcgroty committed:
Creality v4.3.1 (Ender 6) board (#20512)

thisiskeithb committed:
Anet ET4 / ET4P and Anet TFT28 / TFT35 (#20280)

Serhiy-K committed:
STM32F1xx support for TFTGLCD (#20515)

rhapsodyv committed:
Unify FYSETC F6 1.3 / 1.4 (#20507)

Co-authored-by: Scott Lahteine <github@thinkyhead.com>

tpruvot committed:
Fix G28 leveling state, UBL compile (#20499)

spamik committed:
Support RGBW on PCA9632 (#20455)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

spamik committed:
Support RGBW on PCA9632 (#20455)

Co-authored-by: Scott Lahteine <thinkyhead@users.noreply.github.com>

thinkyhead committed:
[cron] Bump distribution date (2020-12-23)

thinkyhead committed:
[cron] Bump distribution date (2020-12-23)

RFBomb committed:
Merge branch 'bugfix-2.0.x' of https://github.com/RFBomb/Marlin_Old into bugfix-2.0.x
